src/zarr/core/dtype/npy/string.py

```diff
@@ -0,0 +1,134 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Self, TypeGuard
+
+import numpy as np
+
+from zarr.core.dtype.npy.common import check_json_str
+from zarr.core.dtype.wrapper import ZDType
+
+if TYPE_CHECKING:
+    from zarr.core.common import JSON, ZarrFormat
+    from zarr.core.dtype.wrapper import TBaseDType
+
+_NUMPY_SUPPORTS_VLEN_STRING = hasattr(np.dtypes, "StringDType")
+
+
+if _NUMPY_SUPPORTS_VLEN_STRING:
+
+    @dataclass(frozen=True, kw_only=True)
+    class VariableLengthString(ZDType[np.dtypes.StringDType, str]):  # type: ignore[type-var]
+        dtype_cls = np.dtypes.StringDType
+        _zarr_v3_name = "numpy.variable_length_utf8"
+
+        @classmethod
+        def _from_dtype_unsafe(cls, dtype: TBaseDType) -> Self:
+            return cls()
+
+        def to_dtype(self) -> np.dtypes.StringDType:
+            return self.dtype_cls()
+
+        @classmethod
+        def check_json(cls, data: JSON, zarr_format: ZarrFormat) -> TypeGuard[JSON]:
+            """
+            Check that the input is a valid JSON representation of a numpy string dtype.
+            """
+            if zarr_format == 2:
+                # TODO: take the entire metadata document in here, and
+                # check the compressors / filters for vlen-utf8
+                # Note that we are checking for the object dtype name.
+                return data == "|O"
+            elif zarr_format == 3:
```

Comment:
note to self: we need to support the string literal `"string"` as an alias here, for backwards compatibility.

cc @rabernat